### PR TITLE
vo_gpu: make screenshots use the GL renderer

### LINKED ISSUE #5498: Screenshots do not preserve GPU-accelerated features such as HDR.
## mpv version and platform

```
mpv 0.28.0-226-g9282a34fbf Copyright © 2000-2018 mpv/MPlayer/mplayer2 projects
 built on Wed Feb  7 02:46:11 CST 2018
ffmpeg library versions:
   libavutil       56.7.100
   libavcodec      58.10.100
   libavformat     58.9.100
   libswscale      5.0.101
   libavfilter     7.11.101
   libswresample   3.0.101
ffmpeg version: N-89971-g8318bf1751
```

```
ffmpeg version N-89971-g8318bf1751 Copyright (c) 2000-2018 the FFmpeg developers
  built with gcc 7.3.0 (GCC)
  configuration: --prefix=/usr --extra-cflags= --extra-ldflags= --disable-rpath --enable-gpl --enable-version3 --enable-nonfree --enable-shared --disable-static --enable-gray --enable-avresample --enable-alsa --enable-avisynth --enable-bzlib --enable-chromaprint --enable-frei0r --enable-gcrypt --enable-gmp --enable-gnutls --enable-iconv --enable-ladspa --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcelt --enable-libcdio --enable-libdc1394 --enable-libfdk-aac --disable-libflite --enable-fontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libiec61883 --enable-libilbc --enable-libjack --enable-libkvazaar --enable-libmodplug --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopencv --enable-libopenh264 --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librsvg --enable-librubberband --enable-librtmp --enable-libshine --enable-libsmbclient --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtesseract --enable-libtheora --disable-libtls --enable-libtwolame --enable-libv4l2 --enable-libvidstab --enable-libvmaf --enable-libvo-amrwbenc --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxavs --enable-libxcb --enable-libxcb-shm --enable-libxcb-xfixes --enable-libxcb-shape --enable-libxvid --enable-libxml2 --enable-libzimg --enable-libzmq --enable-libzvbi --enable-lv2 --enable-lzma --enable-decklink --enable-libmysofa --enable-openal --enable-opencl --enable-opengl --disable-openssl --enable-sndio --enable-sdl2 --enable-xlib --enable-zlib --enable-libdrm --enable-libmfx --enable-nvdec --enable-nvenc --enable-omx --enable-omx-rpi --enable-rkmpp --enable-v4l2-m2m --enable-vaapi --enable-vdpau
  libavutil      56.  7.100 / 56.  7.100
  libavcodec     58. 10.100 / 58. 10.100
  libavformat    58.  9.100 / 58.  9.100
  libavdevice    58.  1.100 / 58.  1.100
  libavfilter     7. 11.101 /  7. 11.101
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  0.101 /  5.  0.101
  libswresample   3.  0.101 /  3.  0.101
  libpostproc    55.  0.100 / 55.  0.100
```

Running on Arch Linux with version 4.14.15-1 of the Linux kernel. The VO driver used is `gpu`, the default with the master branch as of this issue.

## Explanation

Currently, screenshots do not properly output and preserve GPU-accelerated feature(s) such as HDR. Instead, the screenshot that is outputted is an 8-bit version of the 10-bit HDR frame, resulting in a very washed-out image that is not representative of the image displayed in mpv in the slightest. This happens on both the `subtitles` and `video` screenshot options. The `window` option is unaffected because it just literally takes a screenshot of the window, not the video file (at least, I'm assuming that's how it works). The `window` option also scales the image and isn't a proper screenshot of the current frame anyway, so it's not a viable workaround.

I'm unsure if there are other features not present in the outputted image, as I've only tested HDR as of this time. Below are a few examples I tested with "Life of Pi" sample HDR footage, and a ripped UHD BD HDR copy of the animated film "Your Name". All comparison images have been downscaled to 1280x720 for smaller file sizes, and all tone-mapping & related HDR options have been kept to default. The `window` option comparison was not downscaled manually, but by mpv itself by keeping the window at 1280x720.

#### `subtitles` & `video` screenshot options

subtitle option

#### `window` screenshot option

window option

#### `subtitles` & `video` screenshot options

subtitle option

#### `window` screenshot option

window option

---

## Reproduction steps

1. Load provided sample "Life of Pi" HDR footage into mpv.
2. Take a screenshot using the `subtitle` option (default is the `s` key).

## Expected behavior

The outputted screenshot looks like what you're viewing in mpv.

## Actual behavior

The outputted screenshot is incredibly washed-out and entirely unrepresentative of what you are watching. It's even more unrepresentative of non-HDR copies of the same film.

## Log file

```
mpv -v --log-file=out.txt /media/alpha/kiminonawa/old/title01.mkv
```

http://0x0.st/scuL.txt

## Sample files

["Life of Pi" sample HDR footage](http://hdrsamples.com/life-of-pi-4k-uhd-hdr-sample-footage/)

### PR DESCRIPTION
**Pull request by wm4**
Using the GL renderer for color conversion will make sure screenshots
will use the same conversion as normal video rendering.

This drops potentially existing high bit depth and alpha parts from the
final screenshot (i.e. no 16-bit PNG with alpha channel).

Will break for users who use --target-trc and similar options.

I considered creating a new gl_video context, but it could double GPU
memory use, so I didn't.

This uses FBOs instead of glGetTexImage(), because that increases the
chance it could work on GLES (e.g. ANGLE). Untested. No support for the
Vulkan and D3D11 backends yet.

Fixes #5498.

### CODE REVIEW

**Comment by wm4** [comment]
Made corrections. Not sure why the subtitles didn't appear, but I changed it completely anyway. Now the GL renderer will be used to render subtitles (if requested), instead of using the separate OSD software renderer. This has the effect that --blend-subtitle will affect screenshots.

Not sure about that getting darker stuff. A consequence of --hdr-compute and the renderer being reset? Note that this also resets the interpolation state, which apparently can be visible (would probably be preventable, but I won't go into that mess, especially since the renderer is being rewritten by someone else).

**Comment by haasn** [comment]
The only thing that resets the SSBO state is `gl_video_uninit` (which clears it completely). You don't seem to call that from the screenshot path, so this kind of observed behavior would only be explainable by the hdr peak computation if you pause immediately after a transition from a dark scene to a bright scene.

By the way: By absolute necessity, the result of `--hdr-compute` is delayed by a frame. So after `gl_video_uninit` gets called (which does `ra_buf_free` on the `hdr_peak_ssbo`), the first frame will be suboptimal (typically too dark). This also causes a first-frame flicker unless we pre-render a frame offline before we actually start displaying them to the screen. (This would also solve performance spikes and the resulting dropped frames due to first-frame initialization.) But most of the time it goes unnoticed for movies since they like to start on a fade-from-black.

Anyway, can you post screenshots or something?

File: video/out/opengl/ra_gl.c [hunk 1]
```diff
@@ -329,8 +336,11 @@ static struct ra_tex *gl_tex_create(struct ra *ra,
 
     gl_check_error(gl, ra->log, "after creating texture");
 
-    // Even blitting needs an FBO in OpenGL for strange reasons
-    if (tex->params.render_dst || tex->params.blit_src || tex->params.blit_dst) {
+    // Even blitting needs an FBO in OpenGL for strange reasons.
+    // Download is handled by reading from a FBO.
```
Comment (kevmitch) [inline video/out/opengl/ra_gl.c]: an FBO

**Comment by kevmitch** [comment]
The subtitle issue no longer appears.

The image still changes when taking screenshots.

Here are the screenshots. Sorry they're jpeg, since there seem to be unrelated issues with png window screenshots.

I used the sample provided in the original issue:
http://hdrsamples.com/life-of-pi-4k-uhd-hdr-sample-footage/

[log](https://0x0.st/scdM.log)

* Step forward 9 frames 
* Take a [window screenshot](https://0x0.st/scdu.jpg)
* Take an [image screenshot](https://0x0.st/scdS.jpg)
* Take an [image screenshot](https://0x0.st/scdQ.jpg)
* Take a [window screenshot](https://0x0.st/scd1.jpg)

As you can see the window and screenshots are not consistent with themselves or each-other.

If on the other hand, I take only window screenshots, neither the screenshots nor the image on the screen change, as expected.

**Comment by kevmitch** [comment]
With #5510 I can no longer see variations between the screenshots or on the screen.

**Comment by kevmitch** [comment]
@wm4 do you want to add https://github.com/mpv-player/mpv/issues/5240 to the fix list, or keep it open for the underlying problem? Should probably at least reference it.

**Comment by wm4** [comment]
Fixed the typo and readded high bit depth screenshots. (Some weird details change, not sure if anyone cares.)

**Comment by kevmitch** [comment]
nice. thanks.

**Comment by winneon** [comment]
How soon do you expect Vulkan to be supported? More specifically, what changes would need to be made in order to support Vulkan or DirectX 11?

**Pull request merged**